velox/expression/tests/ExprOptimizerTest.cpp

```diff
@@ -0,0 +1,451 @@
+  testExpression(
+      "array_sort(c0, x -> length(x) * abs(-1 * 3 * 2))",
+      "array_sort(c0, x -> length(x) * 6)",
+      ROW({"c0"}, {ARRAY(VARCHAR())}),
+      lambdaTypes,
+      lambdaTypes);
+  lambdaTypes = {DOUBLE(), DOUBLE()};
+  testExpression(
+      "reduce(c0, (89.0E0 + 11.0E0), (s, x) -> s + x * (0.2E0 - 0.1E0), s -> (s < (100.0E0 + 1.0E0)))",
+      "reduce(c0, 100.0E0, (s, x) -> s + x * 0.1E0, s -> (s < 101.0E0))",
+      ROW({"c0"}, {ARRAY(DOUBLE())}),
+      lambdaTypes,
+      lambdaTypes);
+  lambdaTypes = {INTEGER(), INTEGER()};
+  testExpression(
+      "reduce(c0, 100 / 10, (s, x) -> s + x, s -> s + (3 - 2))",
+      "reduce(c0, 10, (s, x) -> s + x, s -> s + 1)",
+      ROW({"c0"}, {ARRAY(INTEGER())}),
+      lambdaTypes,
+      lambdaTypes);
+  lambdaTypes = {SMALLINT(), SMALLINT()};
+  testExpression(
+      "reduce(c0, 1 - 1, (s, x) -> s + x, s -> coalesce(s, abs(2 - 10), 5 * 3) * (5 * 2))",
+      "reduce(c0, 0, (s, x) -> s + x, s -> coalesce(s, 8, 15) * 10)",
+      ROW({"c0"}, {ARRAY(SMALLINT())}),
+      lambdaTypes,
+      lambdaTypes);
+}
+
+/// Test for partial constant folding in the expression tree.
+TEST_F(ExprOptimizerTest, partialConstantFolding) {
+  testExpression(
+      "a = 'z' and b = (1 + 1)",
+      "a = 'z' and b = 2",
+      ROW({"a", "b"}, {VARCHAR(), BIGINT()}));
+  testExpression(
+      "concat(substr(a, 1 + 5 - 2), substr(b, 2 * 3))",
+      "concat(substr(a, 4), substr(b, 6))",
+      ROW({"a", "b"}, {VARCHAR(), VARCHAR()}));
+  testExpression(
+      "a = 'z' and b = (1 + 1) or c = (5 - 1) * 3",
+      "a = 'z' and b = 2 or c = 12",
+      ROW({"a", "b", "c"}, {VARCHAR(), BIGINT(), INTEGER()}));
+
+  auto type = ROW({"a"}, {BIGINT()});
+  testExpression("(123 * 10) + 4 = a", "1234 = a", type);
+  testExpression(
+      "1234 between a and 2000 + 1", "1234 between a and 2001", type);
+
+  type = ROW({"a"}, {VARCHAR()});
+  testExpression(
+      "a='z' and 2 * 3 = 6 / 1 and 2 * 1 = 3 - 2 and abs(-4) = 2 * 2",
+      "a='z' and true and false and true",
+      type);
+  testExpression(
+      "a='z' or 3 - 1 = 4 - 2 or 4 - 2 = 8 / 4 or 8 / 2 = 2 * 3",
+      "a='z' or true or true or false",
+      type);
+}
+
+/// Test combination of expression rewrites and constant folding.
+TEST_F(ExprOptimizerTest, rewritesWithConstantFolding) {
+  auto type = ROW({"c0"}, {ARRAY(VARCHAR())});
+  std::vector<TypePtr> lambdaTypes = {VARCHAR(), VARCHAR()};
+  std::vector<TypePtr> expectedLambdaTypes = {VARCHAR()};
+  testExpression(
+      "array_sort(c0, (x, y) -> if(length(x) < length(y), (2 - 3), if(length(x) > length(y), (-2 + 3), 0 * 3)))",
+      "array_sort(c0, x -> length(x))",
+      type,
+      lambdaTypes,
+      expectedLambdaTypes);
+  testExpression(
+      "array_sort(c0, (x, y) -> if(length(x) < length(y), (-1 * 1), if(length(x) = length(y), abs(0), (2 / 2))))",
+      "array_sort(c0, x -> length(x))",
+      type,
+      lambdaTypes,
+      expectedLambdaTypes);
+  testExpression(
```

Comment:
This test can be removed as well.

Reply:
This is testing rewrite of `array_sort` to `array_sort_desc` and simplification of lambda with a different comparator; so with the two `array_sort` tests present now we are testing combination of rewrite of function call and the simplification of lambda using [comparison checker](https://github.com/facebookincubator/velox/blob/41d3e31ecd0446e6c497e405f42ed980806098a2/velox/functions/lib/SimpleComparisonMatcher.h#L188). Would it be fine to keep these two?